### ISSUE TITLE
[RFC]Qt like File and Path interface for embed resources

### ISSUE THREAD

**Issue by firecoin001** [crystal-lang/crystal#9374]
crystal is good for single binary deployment, but resources such as static assets, locales, database migrations must copy with folder, some shards try to solve this by bake file system, but lack common interface, it 's hard to do integration. 

can we implement a qt like resource interface `:/path/to/file`
then someone can choose backend, such as `Base64`, `ByteArray`, or `tar` and `zip` archive.

then this interface can used by all `File`, `Path`, `Dir`.

eg:

```
File.virtual_filesystem = Base64FileSystem.new()
  .add("../public")
  .add("../locales")
  .add("../db")
```

`I18n.load_paths << ":/locales"`

`Migration.magrations_path = ":/db/migrations"`

`StaticHandler.new(":/public")`

**Comment by jhass** [comment]
There are several shards providing a virtual filesystem. The more popular ones I know are https://github.com/schovi/baked_file_system and https://github.com/busyloop/rucksack.

There's existing related talk at #1649, #2751, #2791, #2886, #5792, #6967.

I think we can close this as a duplicate for #1649 and focus discussion there.

**Comment by straight-shoota** [comment]
Embedding resources is one problem, but how to make them accessible is a different thing. So maybe this could be a separate discussion about integration of virtual filesystems.
This is maybe not super important for stdlib. I think the only component (except for the direct file APIs) that works with file system features is `HTTP::StaticFileHandler`.

**Comment by jhass** [comment]
Well, then the first question is whether stdlib should provide this at all. My personal answer would be no. Crystal should provide a primitive for baking binary data into the binary efficiently, but I think the ideas of how an API ontop of this should look like is too conflicting too easily. 

So I'd love somebody replacing `read_file` with something [`incbin`](https://github.com/crystal-lang/crystal/pull/2791#issuecomment-225299043) based and making it return a reference to a `Bytes` instead of a `StringLiteral`. That should provide enough of a primitive for shards to build nice abstractions on top. Given Crystal's monkey patchability, a shard could even provide an API as suggested in the OP. But I don't see this necessary in stdlib, really.

**Comment by straight-shoota** [comment]
I agree that this doesn't really fit into stdlib. But maybe stdlib could help providing a more pluggable API.
The problem I see with the approach to just prepend paths with a colon is that it only works with a single virtual file system. If an application uses a second virtual file system (for example mounting remote resources), that's already out of limit.
So my idea would be to consider having an abstract API for generic file system access. Stdlib would provide an implementation for the local file system. Other shards could provide implementations for baked files, remote filesystems, and layering them.
The abstraction could also be defined by a shard and the local file system implementation build upon the stdlib features. But having the abstraction in stdlib would ensure compatibility between shards. Otherwise, some shards working with file sytems would only target the stdlib API and it would be hard to make it pluggable.
This is just an undeveloped idea, but it might be worth giving it some though.

**Comment by firecoin001** [comment]
yes we need abstract API !!!

**Comment by mdepot** [comment]
I found this (along with the other issues @jass mentioned above) after searching for a crystal equivalent of golang's embeded binary functionality [here](https://pkg.go.dev/embed).  Perhaps some inspiration can be gleaned from the go implementation.  In go they allow multiple adds like @babyface001 suggests, however the add can be either a local file or a local directory (recursively even).  So instead of just an add function, I might suggest addFile, addDir, and addTree functions to add a single file, a single directory's files, or a full directory tree recursively.

For each of those functions I would also suggest including a second parameter that allows you to also specify a mount point of where in the embedded file system to add to (defaulting to / ).  This would address @straight-shoota's comment about how to deal with a second virtual file system.  It would be handled just like a unix mount point, by adding it into the single embedded file system at some specified mount path.  See go for example.

To address @straight-shoota's other point about mounting remote resources, perhaps we could allow the local path to be specified as a file:// URI instead of just a raw path.  So you could add file://relative/path or file:///absolute/path for local files, and just as easily specify http://example.com/path/to/file (or even sshfs:// or nfs:// URIs in the future.)  These would of course all be accessed at compile time and embedded into the binary during the build.  Since they would be embedded they would by their very nature be read only files, which brings me to my next point.

As to the question of how to support accessing the embedded files at run time via some sort of API, consider this.  If we embrace these files as being read only, then we don't need the robust file system access to them that we would need if they were writable.  Perhaps instead of thinking in terms of accessing the embedded filles as a file system, it might make sense to think in terms of web GET requests.  If we were to extend URI handling to support an embed:// prefix to access the embedded files in the same way that all the existing web functions already support a file:// prefix, then we would already have a robust API designed, and developers would already be familiar with how to use it.  Getting a file from the embedded file system would be coded no differently than getting it via http://.

This has the added benefit that URI's (instead of file paths) would be in use for both the compile/build stage and at runtime, so it would feel more consistent, and by using URI's it allows for future access method expansion on the embedding side later.

While writing an embedded file out to the local disk might take a few more lines of code than a dedicated file based API approach, the tradeoff would be consistency, and developer familiarity with how things are intended to work.  Not to mention it would probably be (relatively) easy to just implement an embed:// URI compared to building a dedicated new file based API around embedding.  With this approach the API would essentially already be designed.
